src/nupic/ntypes/Sdr.hpp

```diff
@@ -582,11 +617,6 @@ class SparseDistributedRepresentation : public Serializable
             if( dimensions[i] != sdr.dimensions[i] )
                 return false;
         }
-        if( hasValue() != sdr.hasValue() )
-            return false;
-        if( !hasValue() )
-            return true;
-
         // Check data
         return std::equal(
             getDense().begin(),
```

Comment:
nit, flat sparse would have been faster to compare (unless dense is somehow used as "main" representation)

Reply:
If this compared flat-sparse data then the flat-sparse would need to be sorted.  This could happen by:
* This method sorting it, which might negates any performance gain
* The API  having a policy of sorted data.  Then the users need to either sort their data or write their code to output sorted data by construction.

> "main" representation

I think that both flat-sparse and dense representations will be used by most SDRs pretty much every compute cycle.  Any sdr which is used as the presynaptic input for Hebbian learning, needs to be dense.  See SpatialPooler.adaptSynapses and TemporalMemory.adaptSegment

Reply:
> flat-sparse would need to be sorted 

good point, the current approach is better.